src/OpenTelemetry.Instrumentation.Wcf/TelemetryPropagationWriter.cs

```diff
@@ -0,0 +1,105 @@
+// <copyright file="TelemetryPropagationWriter.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.ServiceModel.Channels;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Instrumentation.Wcf;
+
+/// <summary>
+/// Pre-defined PropagationWriter callbacks.
+/// </summary>
+public static class TelemetryPropagationWriter
+{
+    /// <summary>
+    /// Writes the values to SOAP message headers. If the message is not a SOAP message it does nothing.
+    /// </summary>
+    /// <param name="request">The outgoing request <see cref="Message"/> (carrier) to write trace context to.</param>
+    /// <param name="name">The header name being written.</param>
+    /// <param name="value">The header value to write.</param>
+    public static void SoapMessageHeaders(Message request, string name, string value)
+    {
+        Guard.ThrowIfNull(request);
+        Guard.ThrowIfNull(name);
+        Guard.ThrowIfNull(value);
+
+        if (request.Version == MessageVersion.None)
+        {
+            return;
+        }
+
+        request.Headers.Add(MessageHeader.CreateHeader(name, "https://www.w3.org/TR/trace-context/", value, false));
```

Comment:
This overload instantiates a new DataContractSerializer instance every time the created header is written to the outgoing SOAP message. You can either cache a static instance of DataContractSerializer created for serializing String and use a different CreateHeader overload which takes XmlObjectSerializer as a parameter (DCS derives from XmlObjectSerializer), or create your own header type that derives from MessageHeader and implement the OnWriteHeaderContents method to call XmlDictionaryWriter.WriteString. The latter option will have slightly better perf, but not by a lot as a string DCS is relatively light weight.